distributed/worker.py

```diff
@@ -1569,26 +1570,52 @@ def add_task(
                 pdb.set_trace()
             raise
 
+    def transitions(self, recommendations):
+        tasks = set()
+        recommendations = recommendations.copy()
+        while recommendations:
+            ts, finish = recommendations.popitem()
+            tasks.add(ts)
+            if isinstance(finish, tuple):
+                finish, kwargs = finish
+            else:
+                kwargs = {}
+            self.log.append((ts.key, ts.state, finish, "recommender"))
+            new_recs = self.transition(ts, finish, **kwargs)
+            recommendations.update(new_recs)
+
+        if self.validate:
+            for ts in tasks:
+                self.validate_task(ts)
+
     def transition(self, ts, finish, **kwargs):
         if ts is None:
             return
         start = ts.state
         if start == finish:
-            return
+            return {}
         func = self._transitions[start, finish]
-        state = func(ts, **kwargs)
-        self.log.append((ts.key, start, state or finish))
-        ts.state = state or finish
-        if self.validate:
-            self.validate_task(ts)
-        self._notify_plugins("transition", ts.key, start, state or finish, **kwargs)
+        recommendations = func(ts, **kwargs)
+        self.log.append((ts.key, start, ts.state))
+        self._notify_plugins("transition", ts.key, start, ts.state, **kwargs)
+        return recommendations
```

Comment:
All changes to `TaskState.state` are handled inside of the respective transition functions now, instead of here.  Instead the individual transition functions return any recommendations to be passed back into the transition loop inside `self.transitions`

(I think the only place the state is mutated outside of a transition function is the dirty hack above in `update_data`)